HW2_3_Internet/code/HW_3.py

```diff
@@ -0,0 +1,368 @@
+# 2. Функцию `long_lasting_function` из примера по понятным причинам запускать не нужно. Достаточно просто убедится, что большие временные интервалы правильно форматируются при отправке сообщения (как в примерах)
+# 3. Допустима реализация логирования, когда логгер полностью перехватывает запись в `stdout` и `stderr` (то есть при выполнении функций печать происходит **только** в файл)
+# 4. В реальной жизни вам не нужно использовать Telegram API при помощи ручных запросов, вместо этого стоит всегда использовать специальные библиотеки Python, реализующие Telegram API, они более высокоуровневые и удобные. В данном задании мы просто учимся работать с API при помощи написания велосипеда.
+# 5. Обязательно прочтите часть конспекта лекции про API перед выполнением задания, так как мы довольно поверхностно затронули это на лекции
+# 
+# **Рекомендуемые к использованию модули:**
+# 1. os
+# 2. sys
+# 3. io
+# 4. datetime
+# 5. requests
+# 6. dotenv
+# 
+# **Запрещённые модули**:
+# 1. Любые библиотеки, реализующие Telegram API в Python (*python-telegram-bot, Telethon, pyrogram, aiogram, telebot* и так далле...)
+# 2. Библиотеки, занимающиеся "перехватыванием" данных из `stdout` и `stderr` (*pytest-capturelog, contextlib, logging*  и так далле...)
+# 
+# 
+# 
+# Результат запуска кода ниже должен быть примерно такой:
+# 
+# ![image.png](attachment:620850d6-6407-4e00-8e43-5f563803d7a5.png)
+# 
+# ![image.png](attachment:65271777-1100-44a5-bdd2-bcd19a6f50a5.png)
+# 
+# ![image.png](attachment:e423686d-5666-4d81-8890-41c3e7b53e43.png)
+
+# Микро-коммент:
+# - Восстановление через `sys.__stdout__/__stderr__` не работает  - сделал через временную переменную. Как я понял, связанно с тем что почему-то у `__...__` указатель тоже перенаправляется ([тред](https://python-list.python.narkive.com/hIt2s03P/why-can-t-i-restore-my-redirected-stdout))
+
+# In[2]:
+
+
+import os
+import sys
+import time
+import datetime
+import io
+import requests
+from dotenv import load_dotenv
+
+
+load_dotenv()
+chat_id = os.environ.get('CHAT_ID')
+bot_token = os.environ.get('TG_API_TOKEN')
+
+def telegram_logger(chat_id):
+    def decorator(func):
+        def inner_function(*args, **kwargs):
+            url = f'https://api.telegram.org/bot{bot_token}/'
+            default_stdout, default_stderr = sys.stdout, sys.stderr
+            buffer_log = io.StringIO()
+            sys.stdout, sys.stderr = buffer_log, buffer_log
+              
+            message = f'💅 Function `{func.__name__}`'
+            try:
+                start = time.time()
```

Comment:
Два `datetime` при вычитании автоматически дают `timedelta`, очень удобно
```suggestion
                start = datetime.now()
```